### PR TITLE
[bugfix] prevent auto complete search outside the given context

### LINKED ISSUE #18: Custom entry lookup searches across multiple contexts
If the configuration file has multiple autocomplete contexts, like this:

```xml
     <autoComplete>
        <context>//w</context>
        <attribute>@lemma</attribute>
        <selection>./text()</selection>
        <request>
            <url>$baseUrl/forms/$selection</url>
        </request>
        <response>
            <transformation>getLemmaOutput.js</transformation>
        </response>
    </autoComplete>
    <autoComplete>
        <context>//rs[@type='place']</context>
        <attribute>@corresp</attribute>  
        <selection>./text()</selection>
        <request>
            <url>$baseUrl/entities/places/$selection</url>
        </request>
        <response>
            <transformation>getPlaceOutput.js</transformation>
        </response>
    </autoComplete>
```

the custom lookup will not always respect the context from which it was called.

Let's say the user calls up a custom look from `<w>` and tries to autocomplete on `@lemma`. Then:

- if the user simply types in the look up window, relying on autocomplete, the lookup will actually use both (or probably all, if there are more than two) URLs from the configuration table:
  
  In the above screenshot, the first two results are from `$baseUrl/forms/$selection` and the second two results are from `$baseUrl/entities/places/$customLookup`
- when the user clicks on the search button, the right call is made (`$baseUrl/forms/$customLookup`), but see also #17 

Please make sure that the custom lookups only use the one lookup url for the given context.

### PR DESCRIPTION
**Pull request by marmoure**
Fixes #18 
This PR will limit the search context and only match the one given

### CODE REVIEW

File: src/main/java/org/humanistika/oxygen/tei/completer/GUI/newSuggestionForm.java [hunk 1]
```diff
@@ -244,7 +251,21 @@ private void fetchjButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-
...
+            attributeExpr = parseXPath(attrXPath);
+        } catch (Exception e) {
+            return;
+        }
+
+
         for (final AutoComplete autoComplete : teiCompleter.getConfiguration().getAutoCompletes()) {
+            //check if attributeExpr addresses a subset of autoCompleteXPaths.attributeXPath
+            final TeiCompleter.AutoCompleteXPaths autoCompleteXPaths = teiCompleter.getXPaths(autoComplete);
+            if (!isSubset(attributeExpr, autoCompleteXPaths.getAttributeXPath())) continue;
```
Comment (adamretter) [inline src/main/java/org/humanistika/oxygen/tei/completer/GUI/newSuggestionForm.java]: To make this less confusing could we rewrite from:

```java
if (!isSubset(attributeExpr, autoCompleteXPaths.getAttributeXPath())) continue;
```

to:
```java
if (isSubset(attributeExpr, autoCompleteXPaths.getAttributeXPath())) {
```

please?

File: src/main/java/org/humanistika/oxygen/tei/completer/GUI/newSuggestionForm.java [hunk 2]
```diff
@@ -391,7 +412,21 @@ protected Object doInBackground() throws Exception {
...
+                attributeExpr = parseXPath(attrXPath);
+            } catch (Exception e) {
+                throw new Exception(e);
+            }
+
+
             for (final AutoComplete autoComplete : teiCompleter.getConfiguration().getAutoCompletes()) {
+                //check if attributeExpr addresses a subset of autoCompleteXPaths.attributeXPath
+                final TeiCompleter.AutoCompleteXPaths autoCompleteXPaths = teiCompleter.getXPaths(autoComplete);
+                if (!isSubset(attributeExpr, autoCompleteXPaths.getAttributeXPath())) continue;
```
Comment (adamretter) [inline src/main/java/org/humanistika/oxygen/tei/completer/GUI/newSuggestionForm.java]: Same comment as above please

File: src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java [hunk 3]
```diff
@@ -235,7 +235,7 @@ private AuthenticationType asClientFactoryAuthenticationType(@Nullable final Aut
         }
     }
 
-    protected class AutoCompleteXPaths {
+    public class AutoCompleteXPaths {
```
Comment (adamretter) [inline src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java:238]: Why is this visibility change needed?

File: src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java [hunk 4]
```diff
@@ -247,7 +247,7 @@ public Expr getAttributeXPath() {
         }
     }
 
-    protected AutoCompleteXPaths getXPaths(final AutoComplete autoComplete) {
+    public AutoCompleteXPaths getXPaths(final AutoComplete autoComplete) {
```
Comment (adamretter) [inline src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java:250]: Why is this visibility change needed?

**Review by adamretter — CHANGES_REQUESTED** [review]
In general this looks good... just some minor changes to improve clarity please, and some questions about visibility of class members

File: src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java (hunk 3 above)
Comment (marmoure) [reply src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java:238]: The GUI source files are not in the same package as `TeiCompleter` packages
```
package org.humanistika.oxygen.tei.completer.GUI;
package org.humanistika.oxygen.tei.completer;
```
Even trying to set it package public doesn't not work in this case

File: src/main/java/org/humanistika/oxygen/tei/completer/GUI/newSuggestionForm.java (hunk 1 above)
Comment (marmoure) [reply src/main/java/org/humanistika/oxygen/tei/completer/GUI/newSuggestionForm.java]: Done

File: src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java (hunk 3 above)
Comment (adamretter) [reply src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java:238]: Okay cool, I see that now, thanks :-)

File: src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java (hunk 4 above)
Comment (adamretter) [reply src/main/java/org/humanistika/oxygen/tei/completer/TeiCompleter.java:250]: Okay cool, I see why we need that now, thanks :-)

**Pull request merged**